<a href="https://colab.research.google.com/github/evan-delasota/483-project1/blob/main/Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 1 Anomaly Detection
## By Evan Delasota, Mathew Enochs, John Tarroza

### Anomaly detection is any process that finds the outliers of a dataset; those items that don't belong.

#### This program compares the results of two algorithms used to detect outliers and prints out the results in a list. The two algorithms used are *Tukey's Fences* and the 68-95-99.7 rule with *standard deviation*.




In [ ]:
from google.colab import files
uploaded = files.upload()

Saving participants.csv to participants.csv


In [ ]:
for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

User uploaded file "participants.csv" with length 1229 bytes


### Week 1 Results

In [ ]:
import csv
import statistics
import pandas as pd

df = pd.read_csv('participants.csv', delimiter=',')
# Change week_name to any other week #s
week_name = 'Week 1'
student_name = "Student Name"
print("Calculating results for ", week_name, "...", sep="")

df_mean = df[week_name].mean()
print("mean:", df_mean, end="\n")

df_median = df[week_name].median()
print("median:", df_median, end="\n")

df_quartiles = df[week_name].quantile([0.25,0.5,0.75]) #Q1, Q2, and Q3
print("week 1 quartiles:\n", df_quartiles, sep="")

q1 = df[week_name].quantile(0.25)
q3 = df[week_name].quantile(0.75)
iqr = q3 - q1
iqr_lower = q1 - (iqr * 1.5)
iqr_upper = q3 + (iqr * 1.5)

iqr_list = []
std_list = []

# Method using Tukey's Fences to detect potential outliers
def tardy_iqr(week_name):
    outlier_list = []

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < iqr_lower or row[week_name] > iqr_upper:
            outlier_list.append(row[student_name])
            
    return outlier_list

iqr_list = tardy_iqr(week_name)
if len(iqr_list) < 1:
    print("No outliers detected")
else:
    print("IQR outlier list:", iqr_list)

# Method using the 68, 95, 99.7 principle involved with standard deviation to 
# detect potential outliers
def tardy_stddev(week_name):
    outlier_list = []
    stddev = statistics.pstdev(df[week_name])
    std_lower = df_mean - (1.5 * stddev)
    std_upper = df_mean + (1.5 * stddev)

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < std_lower or row[week_name] > std_upper:
            outlier_list.append(row[student_name])

    return outlier_list

std_list = tardy_stddev(week_name)
if len(std_list) < 1:
    print("No outliers detected")
else:
    print("STDDEV outlier list:", std_list)

Calculating results for Week 1...
mean: 161.0
median: 175.0
week 1 quartiles:
0.25    174.0
0.50    175.0
0.75    179.0
Name: Week 1, dtype: float64
IQR outlier list: ['Adrian Ellison', 'Tayla Sparrow', 'Owain Emerson', 'Alaya Dickinson']
STDDEV outlier list: ['Adrian Ellison', 'Tayla Sparrow', 'Owain Emerson', 'Alaya Dickinson']


### Week 2 Results

In [ ]:
import csv
import statistics
import pandas as pd

df = pd.read_csv('participants.csv', delimiter=',')
# Change week_name to any other week #s
week_name = 'Week 2'
student_name = "Student Name"
print("Calculating results for ", week_name, "...", sep="")

df_mean = df[week_name].mean()
print("mean:", df_mean, end="\n")

df_median = df[week_name].median()
print("median:", df_median, end="\n")

df_quartiles = df[week_name].quantile([0.25,0.5,0.75]) #Q1, Q2, and Q3
print("week 1 quartiles:\n", df_quartiles, sep="")

q1 = df[week_name].quantile(0.25)
q3 = df[week_name].quantile(0.75)
iqr = q3 - q1
iqr_lower = q1 - (iqr * 1.5)
iqr_upper = q3 + (iqr * 1.5)

iqr_list = []
std_list = []

# Method using Tukey's Fences to detect potential outliers
def tardy_iqr(week_name):
    outlier_list = []

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < iqr_lower or row[week_name] > iqr_upper:
            outlier_list.append(row[student_name])
            
    return outlier_list

iqr_list = tardy_iqr(week_name)
if len(iqr_list) < 1:
    print("No outliers detected")
else:
    print("IQR outlier list:", iqr_list)

# Method using the 68, 95, 99.7 principle involved with standard deviation to 
# detect potential outliers
def tardy_stddev(week_name):
    outlier_list = []
    stddev = statistics.pstdev(df[week_name])
    std_lower = df_mean - (1.5 * stddev)
    std_upper = df_mean + (1.5 * stddev)

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < std_lower or row[week_name] > std_upper:
            outlier_list.append(row[student_name])

    return outlier_list

std_list = tardy_stddev(week_name)
if len(std_list) < 1:
    print("No outliers detected")
else:
    print("STDDEV outlier list:", std_list)

Calculating results for Week 2...
mean: 145.0857142857143
median: 153.0
week 1 quartiles:
0.25    149.0
0.50    153.0
0.75    154.0
Name: Week 2, dtype: float64
IQR outlier list: ['Yasir Fenton', 'Tamara Cottrell', 'Jazmin Foreman', 'Bear Zuniga', 'Miles Lyons', 'Owain Emerson']
STDDEV outlier list: ['Yasir Fenton', 'Tamara Cottrell', 'Miles Lyons', 'Owain Emerson']


### Week 3 Results

In [ ]:
import csv
import statistics
import pandas as pd

df = pd.read_csv('participants.csv', delimiter=',')
# Change week_name to any other week #s
week_name = 'Week 3'
student_name = "Student Name"
print("Calculating results for ", week_name, "...", sep="")

df_mean = df[week_name].mean()
print("mean:", df_mean, end="\n")

df_median = df[week_name].median()
print("median:", df_median, end="\n")

df_quartiles = df[week_name].quantile([0.25,0.5,0.75]) #Q1, Q2, and Q3
print("week 1 quartiles:\n", df_quartiles, sep="")

q1 = df[week_name].quantile(0.25)
q3 = df[week_name].quantile(0.75)
iqr = q3 - q1
iqr_lower = q1 - (iqr * 1.5)
iqr_upper = q3 + (iqr * 1.5)

iqr_list = []
std_list = []

# Method using Tukey's Fences to detect potential outliers
def tardy_iqr(week_name):
    outlier_list = []

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < iqr_lower or row[week_name] > iqr_upper:
            outlier_list.append(row[student_name])
            
    return outlier_list

iqr_list = tardy_iqr(week_name)
if len(iqr_list) < 1:
    print("No outliers detected")
else:
    print("IQR outlier list:", iqr_list)

# Method using the 68, 95, 99.7 principle involved with standard deviation to 
# detect potential outliers
def tardy_stddev(week_name):
    outlier_list = []
    stddev = statistics.pstdev(df[week_name])
    std_lower = df_mean - (1.5 * stddev)
    std_upper = df_mean + (1.5 * stddev)

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < std_lower or row[week_name] > std_upper:
            outlier_list.append(row[student_name])

    return outlier_list

std_list = tardy_stddev(week_name)
if len(std_list) < 1:
    print("No outliers detected")
else:
    print("STDDEV outlier list:", std_list)

Calculating results for Week 3...
mean: 153.8
median: 161.0
week 1 quartiles:
0.25    158.0
0.50    161.0
0.75    163.5
Name: Week 3, dtype: float64
IQR outlier list: ['Adrian Ellison', 'Adeline Jordan', 'Jaye Sweeney', 'Kameron Anthony']
STDDEV outlier list: ['Adrian Ellison', 'Adeline Jordan']


### Week 4 Results

In [ ]:
import csv
import statistics
import pandas as pd

df = pd.read_csv('participants.csv', delimiter=',')
# Change week_name to any other week #s
week_name = 'Week 4'
student_name = "Student Name"
print("Calculating results for ", week_name, "...", sep="")

df_mean = df[week_name].mean()
print("mean:", df_mean, end="\n")

df_median = df[week_name].median()
print("median:", df_median, end="\n")

df_quartiles = df[week_name].quantile([0.25,0.5,0.75]) #Q1, Q2, and Q3
print("week 1 quartiles:\n", df_quartiles, sep="")

q1 = df[week_name].quantile(0.25)
q3 = df[week_name].quantile(0.75)
iqr = q3 - q1
iqr_lower = q1 - (iqr * 1.5)
iqr_upper = q3 + (iqr * 1.5)

iqr_list = []
std_list = []

# Method using Tukey's Fences to detect potential outliers
def tardy_iqr(week_name):
    outlier_list = []

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < iqr_lower or row[week_name] > iqr_upper:
            outlier_list.append(row[student_name])
            
    return outlier_list

iqr_list = tardy_iqr(week_name)
if len(iqr_list) < 1:
    print("No outliers detected")
else:
    print("IQR outlier list:", iqr_list)

# Method using the 68, 95, 99.7 principle involved with standard deviation to 
# detect potential outliers
def tardy_stddev(week_name):
    outlier_list = []
    stddev = statistics.pstdev(df[week_name])
    std_lower = df_mean - (1.5 * stddev)
    std_upper = df_mean + (1.5 * stddev)

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < std_lower or row[week_name] > std_upper:
            outlier_list.append(row[student_name])

    return outlier_list

std_list = tardy_stddev(week_name)
if len(std_list) < 1:
    print("No outliers detected")
else:
    print("STDDEV outlier list:", std_list)

Calculating results for Week 4...
mean: 175.31428571428572
median: 171.0
week 1 quartiles:
0.25    170.0
0.50    171.0
0.75    172.0
Name: Week 4, dtype: float64
IQR outlier list: ['Dora Delacruz', 'Shaquille Wood']
STDDEV outlier list: ['Dora Delacruz']


### Week 5 Results

In [ ]:
import csv
import statistics
import pandas as pd

df = pd.read_csv('participants.csv', delimiter=',')
# Change week_name to any other week #s
week_name = 'Week 5'
student_name = "Student Name"
print("Calculating results for ", week_name, "...", sep="")

df_mean = df[week_name].mean()
print("mean:", df_mean, end="\n")

df_median = df[week_name].median()
print("median:", df_median, end="\n")

df_quartiles = df[week_name].quantile([0.25,0.5,0.75]) #Q1, Q2, and Q3
print("week 1 quartiles:\n", df_quartiles, sep="")

q1 = df[week_name].quantile(0.25)
q3 = df[week_name].quantile(0.75)
iqr = q3 - q1
iqr_lower = q1 - (iqr * 1.5)
iqr_upper = q3 + (iqr * 1.5)

iqr_list = []
std_list = []

# Method using Tukey's Fences to detect potential outliers
def tardy_iqr(week_name):
    outlier_list = []

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < iqr_lower or row[week_name] > iqr_upper:
            outlier_list.append(row[student_name])
            
    return outlier_list

iqr_list = tardy_iqr(week_name)
if len(iqr_list) < 1:
    print("No outliers detected")
else:
    print("IQR outlier list:", iqr_list)

# Method using the 68, 95, 99.7 principle involved with standard deviation to 
# detect potential outliers
def tardy_stddev(week_name):
    outlier_list = []
    stddev = statistics.pstdev(df[week_name])
    std_lower = df_mean - (1.5 * stddev)
    std_upper = df_mean + (1.5 * stddev)

    for index, row in df.iterrows():
        # if value is out of range, append to outlier list
        if row[week_name] < std_lower or row[week_name] > std_upper:
            outlier_list.append(row[student_name])

    return outlier_list

std_list = tardy_stddev(week_name)
if len(std_list) < 1:
    print("No outliers detected")
else:
    print("STDDEV outlier list:", std_list)

Calculating results for Week 5...
mean: 161.8
median: 171.0
week 1 quartiles:
0.25    168.0
0.50    171.0
0.75    174.5
Name: Week 5, dtype: float64
IQR outlier list: ['Jazmin Foreman', 'Sanjay Edwards', 'Alfie-James Pierce', 'Adeline Jordan', 'Saffa Brook']
STDDEV outlier list: ['Jazmin Foreman', 'Sanjay Edwards', 'Alfie-James Pierce']


## Do your results agree with experiment (4)?
- Yes, they do!

## Explanation
- We created a Python code that loads in the data set and examine the contents of the students' attendance from week 1 - 5. Our code found the quartiles, used the Tukey's fences, and standard deviation to find the anomalies in the dataset. Student's that are listed as outliers are the anomalies. The results are:
  ### Calculating results for Week 1...
  - mean: 161.0
  - median: 175.0
  - week 1 quartiles:
    - 0.25:    174.0
    - 0.50:    175.0
    - 0.75:    179.0
  - Name: Week 1, dtype: float64
  - IQR outlier list: ['Adrian Ellison', 'Tayla Sparrow', 'Owain Emerson', 'Alaya Dickinson']
  - STDDEV outlier list: ['Adrian Ellison', 'Tayla Sparrow', 'Owain Emerson', 'Alaya Dickinson']

  ### Calculating results for Week 2...
  - mean: 145.0857142857143
  - median: 153.0
  - week 1 quartiles:
    - 0.25:    149.0
    - 0.50:   153.0
    - 0.75:    154.0
  - Name: Week 2, dtype: float64
  - IQR outlier list: ['Yasir Fenton', 'Tamara Cottrell', 'Jazmin Foreman', 'Bear Zuniga', 'Miles Lyons', 'Owain Emerson']
  - STDDEV outlier list: ['Yasir Fenton', 'Tamara Cottrell', 'Miles Lyons', 'Owain Emerson']

  ### Calculating results for Week 3...
  - mean: 153.8
  - median: 161.0
  - week 1 quartiles:
    - 0.25:    158.0
    - 0.50:    161.0
    - 0.75:    163.5
  - Name: Week 3, dtype: float64
  - IQR outlier list: ['Adrian Ellison', 'Adeline Jordan', 'Jaye Sweeney', 'Kameron Anthony']
  - STDDEV outlier list: ['Adrian Ellison', 'Adeline Jordan']

  ### Calculating results for Week 4...
  - mean: 175.31428571428572
  - median: 171.0
  - week 1 quartiles:
    - 0.25:    170.0
    - 0.50:   171.0
    - 0.75:    172.0
  - Name: Week 4, dtype: float64
  - IQR outlier list: ['Dora Delacruz', 'Shaquille Wood']
  - STDDEV outlier list: ['Dora Delacruz']

  ### Calculating results for Week 5...
  - mean: 161.8
  - median: 171.0
  - week 1 quartiles:
    - 0.25:    168.0
    - 0.50:    171.0
    - 0.75:    174.5
  - Name: Week 5, dtype: float64
  - IQR outlier list: ['Jazmin Foreman', 'Sanjay Edwards', 'Alfie-James Pierce', 'Adeline Jordan', 'Saffa Brook']
  - STDDEV outlier list: ['Jazmin Foreman', 'Sanjay Edwards', 'Alfie-James Pierce']

